validator/pandora/service.go

```diff
@@ -0,0 +1,214 @@
+	// Resume if eth1 node is synced.
+	if synced {
+		s.connected = true
+		s.runError = nil
+		log.WithFields(logrus.Fields{"endpoint": logutil.MaskCredentialsLogging(s.endpoint),
+		}).Info("Connected to pandora chain")
+		return
+	}
+	if errSynced != nil {
+		s.runError = errSynced
+		log.WithError(errSynced).Error("Could not check sync status of pandora chain")
+	}
+
+	// Use a custom logger to only log errors
+	// once in  a while.
+	logCounter := 0
+	errorLogger := func(err error, msg string) {
+		if logCounter > logThreshold {
+			log.WithError(err).Error(msg)
+			logCounter = 0
+		}
+		logCounter++
+	}
+
+	ticker := time.NewTicker(dialInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			log.Debugf("Trying to dial endpoint: %s", s.endpoint)
+			synced, errSynced := s.isPandoraNodeSynced()
+			if errSynced != nil {
+				errorLogger(errSynced, "Could not check sync status of pandora chain")
+				s.runError = errSynced
+				continue
+			}
+			if synced {
+				s.connected = true
+				s.runError = nil
+				log.WithFields(logrus.Fields{
+					"endpoint": logutil.MaskCredentialsLogging(s.endpoint)}).
+					Info("Connected to pandora chain")
+				return
+			}
+			s.runError = errNotSynced
+		case <-s.ctx.Done():
+			log.Debug("Received cancelled context, closing existing pandora client service")
+			return
+		}
+	}
+}
+
+// GetPandoraBlock method calls pandora client's `eth_getWork` api and decode header and extra data fields
+// This methods returns eth1Types.Header and ExtraData
+func (s *Service) GetWork(ctx context.Context) (*eth1Types.Header, *ExtraData, error) {
+	if !s.connected {
+		log.WithError(ConnectionError).Error("Pandora chain is not connected")
+		return nil, nil, ConnectionError
+	}
+
+	response, err := s.pandoraClient.GetWork(ctx)
+	if err != nil {
+		log.WithError(err).Error("Pandora block preparation failed")
+		return nil, nil, err
+	}
+	header := response.Header
+	var extraData ExtraData
+	if err := rlp.DecodeBytes(header.Extra, &extraData); err != nil {
+		return nil, nil, errors.Wrap(err, "Failed to decode extra data fields")
+	}
+	return header, &extraData, nil
+}
+
+// SubmitWork method calls pandora client's `eth_submitWork` api
+// This method returns a boolean status
+func (s *Service) SubmitWork(ctx context.Context, blockNonce uint64,
+	headerHash common.Hash, sig [32]byte) (bool, error) {
+
+	if !s.connected {
+		log.WithError(ConnectionError).Error("Pandora chain is not connected")
+		return false, ConnectionError
+	}
+
+	status, err := s.pandoraClient.SubmitWork(ctx, blockNonce, headerHash, sig)
+	if err != nil || !status {
+		log.WithError(err).Error("Work submission failed")
+		return false, err
+	}
+	return status, nil
+}
+
+// checks if the pandora node is healthy and ready to serve before
+// fetching data from  it.
+func (s *Service) isPandoraNodeSynced() (bool, error) {
+	syncProg, err := s.pandoraClient.SyncProgress(s.ctx)
+	if err != nil {
```

Comment:
you do not need to have if, `return syncProg != nil, err`

Reply:
okay and done.